spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -152,20 +153,30 @@
       this.localityPreferred = false;
     }
 
-    this.schema = table.schema();
+    this.schema = snapshotSchema();
     this.caseSensitive = caseSensitive;
     this.batchSize = options.get(SparkReadOptions.VECTORIZATION_BATCH_SIZE).map(Integer::parseInt).orElseGet(() ->
         PropertyUtil.propertyAsInt(table.properties(),
           TableProperties.PARQUET_BATCH_SIZE, TableProperties.PARQUET_BATCH_SIZE_DEFAULT));
   }
 
+  protected Schema snapshotSchema() {
+    if (snapshotId != null && table instanceof BaseTable) {
```

Comment:
I think this should use the snapshot's schema ID, like this:

```java
if (snapshotId != null) {
  Schema snapshotSchema = table.schemas().get(table.snapshot(snapshotId).schemaId());
  if (snapshotSchema != null) {
    return snapshotSchema;
  }

  // TODO: recover the schema using a util class
  return table.schema();
} else {
  return table.schema();
}
```

Reply:
Yes, I'll follow your suggestion so we don't need to change `BaseTable`.
Also, as you point out below, this same code appears in several places. I'll put it in a util method.